File: companies-leetcode/google/arrays-and-strings/add-bold-tag-in-string.js
/*
Given a string s and a list of strings dict, you need to add a closed pair of bold tag <b> and </b> to wrap the substrings in s that exist in dict. If two such substrings overlap, you need to wrap them together by only one pair of closed bold tag. Also, if two substrings wrapped by bold tags are consecutive, you need to combine them.
Example 1:
Input: 
s = "abcxyz123"
dict = ["abc","123"]
Output:
"<b>abc</b>xyz<b>123</b>"
Example 2:
Input: 
s = "aaabbcc"
dict = ["aaa","aab","bc"]
Output:
"<b>aaabbc</b>c"
Note:
The given dict won't contain duplicates, and its length won't exceed 100.
All the strings in input have length in range [1, 1000].
*/
// 100th percentile 
/**
 * @param {string} s
 * @param {string[]} dict
 * @return {string}
 */
var addBoldTag = function(s, dict) {
    if (!s || !dict) return "";
    
    let bolded = new Array(s.length).fill(false);
    
    // Identify bolded characters
    for (let i = 0; i < dict.length; i++) {
        let nxtTag = dict[i];
        let start = s.indexOf(nxtTag);
        while (start !== -1) {
            bolded.fill(true, start, start+nxtTag.length)
            start = s.indexOf(nxtTag, start+1);
        }
    }
    
    // Construct final string
    let res = "";
    let bold = bolded[0];
    let nxtString = bold ? "<b>" + s[0] : s[0];
    
    for (let k = 1; k < s.length; k++) {
        // Completed a bolded / unbolded section
        if (bolded[k] !== bolded[k-1]) {
            if (bold) {
                res += nxtString + "</b>";
                bold = false;
                nxtString = s[k];
            } else {
                res += nxtString;
                bold = true;
                nxtString = "<b>" + s[k];
            }
        } 
        // Continue constructing the string
        else {
            nxtString += s[k];
        }
        // Last character
        if (k === s.length-1) {
            res = bold ? res + nxtString + "</b>" : res + nxtString;
        }
    }
    
    return res;
};

File: companies-leetcode/google/arrays-and-strings/add-bold-tag-in-string.py
"""
Given a string s and a list of strings dict, you need to add a closed pair of bold tag <b> and </b> to wrap the substrings in s that exist in dict. If two such substrings overlap, you need to wrap them together by only one pair of closed bold tag. Also, if two substrings wrapped by bold tags are consecutive, you need to combine them.
Example 1:
Input: 
s = "abcxyz123"
dict = ["abc","123"]
Output:
"<b>abc</b>xyz<b>123</b>"
Example 2:
Input: 
s = "aaabbcc"
dict = ["aaa","aab","bc"]
Output:
"<b>aaabbc</b>c"
Note:
The given dict won't contain duplicates, and its length won't exceed 100.
All the strings in input have length in range [1, 1000].
"""
# 81st percentile
class Solution:
    def addBoldTag(self, s: str, dict: List[str]) -> str:
        if not dict: return s
        if not s: return ""
        
        bolded = [ False for _ in s ]
        
        # Identify characters to be bolded
        for tag in dict:
            start = s.find(tag)
            while start != -1:
                for i in range(start, start + len(tag)):
                    bolded[i] = True
                # bolded[start:(start+len(tag))] = [True]*(start+len(tag))
                start = s.find(tag, start+1)
        
        # Construct output string
        bold = bolded[0]
        nxtString = "<b>" + s[0] if bold else s[0]
        res = ""
        
        for i in range(1, len(bolded)):                    
            # Finished a bold / non-bold section
            if bolded[i]!=bolded[i-1]:
                if bold:
                    res += nxtString + "</b>"
                    bold = False
                    nxtString = s[i]
                else:
                    res += nxtString
                    bold = True
                    nxtString = "<b>" + s[i]
                    
            # Continue constructing a section
            else:
                nxtString += s[i]
                        # End of string
            if i==len(s)-1:
                res = res + nxtString + "</b>" if bold else res + nxtString

        return res